### DISCUSSION TITLE
Map before write and after read?

### DISCUSSION THREAD

**Discussion by m1212e** [dexie/Dexie.js#1456] [Q&A] [answered]
Hi,

is there a possibility to automatically map objects? I'm looking for something you can register once and then you can just use the default methods just as usual but internally the data will be mapped before every read/write. This would be pretty useful, since the data stored doesn't always match the object you want to consume later on. Currently I'm calling a helper function every time before storing and after reading data. It would be super elegant to just let dexie handle the map automatically without having to worry about mapping manually. I suspect it should be possible with hooks/ Dexie.use() but I couldn't get it to work. Could someone maybe provide a quick breakdown of what is required to achieve this?
Thanks!

**Comment by dfahlander** [comment, marked as answer]
It's possible both via hooks and middleware.

Read mapping is easiest implemented via hook('reading'). Writes via a middleware that overrides mutate. It's a bit verbose to implement but you only need to copy the example from https://dexie.org/docs/Dexie/Dexie.use(), check if req.type === 'put' || req.type === 'add' and if the table is the target one, then replace myRequest.values = req.values.map(yourMapper)

**Reply by Zetanova** [reply]
Is this still valid?

It seams a but much to hook up a custom mapper/serializer into dexie.

I have the simple use case to workaround this issue https://github.com/bufbuild/protobuf-es/issues/508
by looking on each field value if a "toJson"/"toBinary" method exist and replacing the field value with the result
and in reverse manner "fromJson"/"fromBinary" but for this a Schema would be required

My current findings are:
to read: https://github.com/dexie/Dexie.js/issues/1180
to write:  https://dexie.org/docs/Dexie/Dexie.use()

Then I found something promise looking:
`TableSchema.mappedClass` 
`TableSchema.readHook` 
but only little documentation
